### PR TITLE
[docs][proposal] Magma QoS Policy and Federation API Cleanup

### PR DESCRIPTION
**Pull request by andreilee**
Signed-off-by: Andrei Lee <andreilee@fb.com>

## Summary

Proposal for QoS Enforcement

## Test Plan

N/A

### CODE REVIEW

File: docs/qos_enforcement.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+The orc8r policydb service should be updated to stream to the gateway all the mappings between basenames, policies, QoS profiles, APNs, data plans and subscribers.
+
+Gateway policydb service needs to be updated to make RAR type calls to sessiond when updates happen to the policy associations or to policies. To know when this is required, policy information and associations are streamed from orc8r, and policydb should also query sessiond for the relevant info.
+
+
+### Session Manager Changes
+
+The session manager should provide grpc methods to retrieve the currently active sessions. With each session, the IMSI, APN, and active policies should be provided. This surface will be used by policydb.
+
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: is this needed? sessiond already sends create/terminate session requests to policydb. So policydb can track which sessions are active and push any modifications on policy rules.

File: docs/qos_enforcement.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+**DELETE  ****/****lte****/{****network_id****}/****apns****/{****apn_name****}**
+
+
+```
+
+
+
+### API Entity Definitions
+
+Entity definitions for QoS, APN, and subscribers will change. The final state is summarized here:
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: Let's add back the GBR field. This will make it consistent with the spec. Even if AGW is not enforcing it, RAN can still enforce it (particularly important for the wireless scheduler) as control plane carries this information.

Comment (ulaskozat) [reply docs/qos_enforcement.md]: also note that PCRF also utilizes these fields in corresponding AVPs.

File: docs/qos_enforcement.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  ...
+  qos: array[qos_profile]
+```
+
+Multiple QoS objects will be allowed to be associated to each policy object. To implement this, its *qos* field will be changed from the current *flow_qos* type to the new QoS object type, as defined above.
+
+
+### Enforcement with Pipelined Service
+
+Changes in pipelined are required to enforce QoS via AMBR across the bearer level, APN level, and subscriber level.
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: Also define a To Do (lower priority):
- GBR enforcement (P3)
- Load/KPI reporting (to be used for admission control and session lifecycle management by sessiond) (P2)

File: docs/qos_enforcement.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+
+### Policydb Changes with Streaming
+
+The orc8r policydb service should be updated to stream to the gateway all the mappings between basenames, policies, QoS profiles, APNs, data plans and subscribers.
+
+Gateway policydb service needs to be updated to make RAR type calls to sessiond when updates happen to the policy associations or to policies. To know when this is required, policy information and associations are streamed from orc8r, and policydb should also query sessiond for the relevant info.
+
+
+### Session Manager Changes
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: More changes:
- dedicated bearer LCM (P1) 
- admission control and preemption (P2)

File: docs/qos_enforcement.md (hunk 1 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: Having `sessiond` be the source of truth here can keep `policydb` stateless. On principle also, I'm a strong believer of making sure we don't have duplicated information.

File: docs/qos_enforcement.md (hunk 2 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: Sure I'll add it back in

File: docs/qos_enforcement.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the default QoS object.
+
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_service_type:
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: Service is a very confusing term to use here. we are allocating bunch of logical and physical resources to APN. I would use `apn_resources` instead.

File: docs/qos_enforcement.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  supported_qos_profiles: array[qos_profile]
+  # service_type should match the name of a specified ApnServiceType
+  service_type: string
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: change `service_type` to `resource_label`?

Comment (ulaskozat) [reply docs/qos_enforcement.md]: Actually this does not make sense to me to define this as part of APN configuration. This is an AGW level resource. Why don't we include this as part of gateway configuration.

File: docs/qos_enforcement.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  supported_qos_profiles: array[qos_profile]
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: shall we make default QoS explicit?
```
apn_configuration:
  default_qos_profile: qos_profile
  dedicated_qos_services: array[qos_profile]
```

**Comment by ulaskozat** [comment]
@andreilee @amarpad @xjtian @pshelar, do we want IP block, vlan, interface, etc. stuff that we will provision for an APN at network level or AGW level? I thought our goal is to have APN->vlan or APN->nic mapping  at the AGW level. The current definitions make it a network level affair.

If I recall correctly, the idea is to create an indirection where APN is passed a network resource name and at the gateway level define the mappings from network resource name to actual resource definitions. When streaming down to subscriberdb for a specific AGW, streamer can pull the AGW level mapping and populate the information.

File: docs/qos_enforcement.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+One can be specified as the APN-level default policy. This is checked by verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the default QoS object.
+
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_service_type:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN service type
+  allocated_ip_range: string
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: allocated_ip_range -> ipv4_block

File: docs/qos_enforcement.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  # Specify either VLAN or the interface name for backhaul
+  vlan_id: string
+  interface: string
+```
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: shall we add `ipv6_prefix: string`

File: docs/qos_enforcement.md (hunk 6 above)
Comment (ulaskozat) [reply docs/qos_enforcement.md]: strike my comment, actually you are doing it right:  
AGW -> services
APN -> service
service -> actual network resource

File: docs/qos_enforcement.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+
+APN Service Type
+
+GET     /lte/{network_id}/gateways/{gateway_id}/service_types
+POST    /lte/{network_id}/gateways/{gateway_id}/service_types
+
+GET     /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
+PUT     /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
+DELETE  /lte/{network_id}/gateways/{gateway_id}/service_types/{service_type}
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: `service_types` -> `resource_labels`
`service-type` -> `resource_label`

File: docs/qos_enforcement.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+Multiple QoS objects will be allowed to be associated to each policy object. To implement this, its *qos* field will be changed from the current *flow_qos* type to the new QoS object type, as defined above.
+
+
+### Pipelined Changes for Enforcement
+
+At a high level, changes in `pipelined` are required to support the following:
+* (P1) Enforce QoS via AMBR across the bearer level, APN level, and subscriber level
+* (P2) Load/KPI reporting (to be used for admission control and session lifecycle management by sessiond)
+* (P3) GBR enforcement
+
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: any changes to proto files?

File: docs/qos_enforcement.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+of a single subscriber, one for each APN. Across these different sessions, each 
+APN will have a different QoS for its default bearer. To enable this behavior, 
+there should be the capability to assign a QoS profile to an APN for the 
+default bearer.
+
+It is also possible that a network operator would desire to have backhaul 
+specified differently for each APN even when served from the same gateway. 
+This requirement can arise naturally from prioritizing one service type, and 
+thus APN, over another. 
+We will define the `ApnServiceType` as a configuration specifying how backhaul
```
Comment (amarpad) [inline docs/qos_enforcement.md]: @ulaskozat was suggesting a name with Resource somewhere. I am ok with ServiceType too.

I would reword how backhaul is setup for an APN in the gateway to how SGi interface(s) are setup for APN in the gateway.

File: docs/qos_enforcement.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+4. Assignment of QoS profiles to APN entities (to denote what QoS levels are supported by the APN)
+5. Assignment of APN entities to subscribers (to denote which APNs a subscriber has access to)
+6. Assignment of ApnServiceType entities to gateways
+
+**Federated LTE**
+
+In federated use cases, we have slightly different requirements. APN entities 
+do not need to be configured as they are provided by the federation network.
+
+We also wish to support cases where an HSS is not provided, and our
```
Comment (amarpad) [inline docs/qos_enforcement.md]: For the case where our subscriberdb is used we will need to allow for APN entities to be specified by us. i.e. only when HSS is federated we disable our APN configuration.

File: docs/qos_enforcement.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+In federated use cases, we have slightly different requirements. APN entities 
+do not need to be configured as they are provided by the federation network.
+
+We also wish to support cases where an HSS is not provided, and our 
+`subscriberdb` service is used, but where the PCRF is federated, 
+and so our `policydb` service is disabled.
+
+This adds the following requirement:
+
+7. Disabled configuration of APN entities for federated LTE networks
```
Comment (amarpad) [inline docs/qos_enforcement.md]: when HSS federation is enabled.

Also, there is another thread with @uri200 and @themarwhal where we should be able to disable Gx or Gy selectively.

File: docs/qos_enforcement.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_service_type:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN service type
+  allocated_ip_range: string
+  # Specify either VLAN or the interface name for backhaul
```
Comment (amarpad) [inline docs/qos_enforcement.md]: replace backhaul with SGi as that is probably the more commonly used term

File: docs/qos_enforcement.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+  vlan_id: string
+  interface: string
```
Comment (amarpad) [inline docs/qos_enforcement.md]: I think I would use uplink_interfaces and add a comment that the uplink_bridge (check name in installer code), needs to be configured with the uplink_interface for this configuration to be valid.

File: docs/qos_enforcement.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_service_type:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN service type
+  allocated_ip_range: string
+  # Specify either VLAN or the interface name for backhaul
+  vlan_id: string
```
Comment (amarpad) [inline docs/qos_enforcement.md]: vlan to use when communicating over the SGi interface

**Review by amarpad — DISMISSED** [review]

File: docs/qos_enforcement.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+thus APN, over another. 
+We will define the `ApnServiceType` as a configuration specifying how backhaul 
+is set up for an APN in a gateway. 
+
+To enable the above, we have the following list of new API requirements for Magma:
+
+1. Configuration of QoS profiles
+2. Configuration of APN entities
+3. Configuration of ApnServiceType entities
+3. Assignment of QoS profiles to subscribers (to denote what QoS levels are allowed for the subscriber)
```
Comment (xjtian) [inline docs/qos_enforcement.md]: Is this a workflow step that we need to do in the API, or is this just something we get from the association of subscriber -> APNs?

File: docs/qos_enforcement.md (hunk 7 above)
Comment (xjtian) [reply docs/qos_enforcement.md]: agreed

**Review by xjtian — DISMISSED** [review]
Dismiss this review status when you finish updating the doc w/ the existing feedback

File: docs/qos_enforcement.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+The `policydb` service across the orc8r and gateway requires changes to support QoS enforcement:
+* (P1) Orc8r `policydb` service should be updated to stream to the gateway all the mappings between basenames, policies, QoS profiles, APNs, data plans and subscribers
+* (P1) Gateway `policydb` service should make RAR calls to `sessiond` when updates happen to the policy associations or to policies. To determine when required, policy info and associations are streamed from orc8r, and `policydb `should also query `sessiond` for the relevant info.
+
+
+### Session Manager Changes
+
+The session manager will be changed to support the following:
+* (P1) Provide gRPC methods to retrieve the currently active sessions, including the IMSI, APN, and active policies
+* (P1) Each APN will be assigned a QoS profile to be used for the default bearer
```
Comment (themarwhal) [inline docs/qos_enforcement.md]: Will `policyDB` handle the logic of which QoS profile is applied? What needs to be maintained on `SessionD`s side?

File: docs/qos_enforcement.md (hunk 14 above)
Comment (uri200) [reply docs/qos_enforcement.md]: yes, once that is in place and OMNI rules support charging, we will be able to have `Gy without Gx` or `only Gx` networks. But per I can read here I think this won't be impacted by this new feature.

File: docs/qos_enforcement.md (hunk 11 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: I think that would be for either Nick or Pravin to look into

File: docs/qos_enforcement.md (hunk 18 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: We'd get it from the assoc. Let me remove this line

File: docs/qos_enforcement.md (hunk 19 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: I think that information should be given from policydb in the `CreateSessionResponse` @ulaskozat

File: docs/qos_enforcement.md [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+  # this APN resource
+  ipv4_block: string 
+  # Specify either VLAN or the interface name for SGi
+  # VLAN to use when communicating over SGi interface
+  vlan_id: string
+  # uplink_bridge must be configured with the uplink interface for this value 
+  # to be valid.
+  uplink_interfaces: string
+  # Prefix for ipv6 allocation
+  ipv6_prefix: string
```
Comment (amarpad) [inline docs/qos_enforcement.md]: Do we model an IPAddress object with both v6 and v4 also what happens during attach we return both the v6 and v4 address?

Comment (ulaskozat) [reply docs/qos_enforcement.md]: We need to include PDNType in the APN Configuration. If it is set as IPV4V6, then both IP addresses are assigned and returned (to be completed in Q3) during attach.

File: docs/qos_enforcement.md [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+matures, we wish to enable QoS enforcement for multiple sessions 
+of a single subscriber, one for each APN. Across these different sessions, each 
+APN will have a different QoS for its default bearer. To enable this behavior, 
+there should be the capability to assign a QoS profile to an APN for the 
+default bearer.
+
+It is also possible that a network operator would desire to have backhaul 
+specified differently for each APN even when served from the same gateway. 
+This requirement can arise naturally from prioritizing one service type, and 
+thus APN, over another.
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: This comment is not clear to me. what does it mean to prioritize one APN over the other?

File: docs/qos_enforcement.md [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by 
+verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the 
+default QoS object.
+
+Additional QoS objects can be associated to the APN for additional resources 
+beyond the default policy, both physical and logical.
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: This may be confused with the apn_resource object defined below. Rephrasing to "Additional QoS objects can be associated with the APN to deliver dedicated QoS services for susbcribers' traffic".

File: docs/qos_enforcement.md [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+
+```
+policy_rule:
+  id: string
+  flow_list: array[flow_description]
+  priority: uint32
+  rating_group: uint32
+  monitoring_key: string
+  ...
+  qos: array[qos_profile]
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: There should be only one qos_profile for a given policy rule. Otherwise it won't be clear which QoS rule applies to which flow description.

File: docs/qos_enforcement.md [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  default_qos_profile: string
+  supported_qos_profiles: array[string]
+  # resource_label should match the name of a specified ApnResource
+  resource_label: string
```
Comment (ulaskozat) [inline docs/qos_enforcement.md]: we should add a mandatory `pdn_type` field as enum. see subscriberdb.proto for the possible types. IPV6 and IPV4V6 are being developed in Q3.

File: docs/qos_enforcement.md (hunk 21 above)
Comment (andreilee) [reply docs/qos_enforcement.md]: I'll just remove this sentence it's not really adding anything

**Comment by talkhasib** [comment]
Consider changing file location from root "docs" folder

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+
+```
+apn_configuration:
+  default_qos_profile: string
+  supported_qos_profiles: array[string]
+  # resource_label should match the name of a specified ApnResource
+  resource_label: string
+  # Unsupported as this time
+  # To be mandatory
+  pdn_type: enum[IPV4, IPV6, IPV4_IPV6]
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: IPV4_IPv6 is not clear. Let's change this to either as IPV4V6 inline with `PDNType` definition in subscriberdb.proto or as IPV4_AND_IPV6.

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+
+One QoS objects can be associated to each policy object.
+To implement this, its *qos* field will be changed from the current *flow_qos* 
+type to an ID specifier for the new QoS object type, as defined above.
+
+
+### Pipelined Changes for Enforcement
+
+At a high level, changes in `pipelined` are required to support the following:
+* (P1) Enforce QoS via AMBR across the bearer level, APN level, and subscriber
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: subscriber level is not needed as UE_AMBR is enforced at RAN side.

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 27]
```diff
@@ -0,0 +1,333 @@
...
+required, policy info and associations are streamed from orc8r, and `policydb` 
+should also query `sessiond` for the relevant info.
+
+
+### Session Manager Changes
+
+The session manager will be changed to support the following:
+* (P1) Provide gRPC methods to retrieve the currently active sessions, 
+including the IMSI, APN, and active policies
+* (P1) Each APN will be assigned a QoS profile to be used for the default bearer
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: session already has a default QoS profile coming from the MME side and this is stored as part of the session context.

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 28]
```diff
@@ -0,0 +1,333 @@
...
+* (P1) Gateway `policydb` service should make RAR calls to `sessiond` when 
+updates happen to the policy associations or to policies. To determine when 
+required, policy info and associations are streamed from orc8r, and `policydb` 
+should also query `sessiond` for the relevant info.
+
+
+### Session Manager Changes
+
+The session manager will be changed to support the following:
+* (P1) Provide gRPC methods to retrieve the currently active sessions,
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: Let's avoid this and instead provide a set interface that pipelined can use to set new policy rules at sessiond.

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 29]
```diff
@@ -0,0 +1,333 @@
...
+The `policydb` service across the orc8r and gateway requires changes to support 
+QoS enforcement:
+* (P1) Rename `relay_enabled` flag to `gx_gy_relay_enabled`
+* (P1) Orc8r `policydb` service should be updated to stream to the gateway 
+all the mappings between basenames, policies, QoS profiles, APNs, data plans 
+and subscribers
+* (P1) Gateway `policydb` service should make RAR calls to `sessiond` when 
+updates happen to the policy associations or to policies. To determine when 
+required, policy info and associations are streamed from orc8r, and `policydb` 
+should also query `sessiond` for the relevant info.
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: Let's switch to a model where policydb uses a set RPC call to sessiond to update policy rules. APN and subscriber info should be present if this is specific to a particular session.

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 30]
```diff
@@ -0,0 +1,333 @@
...
+
+
+### Policydb Changes with Streaming
+
+The `policydb` service across the orc8r and gateway requires changes to support 
+QoS enforcement:
+* (P1) Rename `relay_enabled` flag to `gx_gy_relay_enabled`
+* (P1) Orc8r `policydb` service should be updated to stream to the gateway 
+all the mappings between basenames, policies, QoS profiles, APNs, data plans 
+and subscribers
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md:300]: default QoS policy is streamed down to subscriberdb as part of APN configuration.

**Comment by hcgatewood** [comment]
# Orc8r clarifying questions

I'm starting on #2257, which should be the bulk of the orc8r work. I have some clarifying questions, below, as I'm digging into this.

## Context

My main focus here is clearly specifying the Swagger models, because I'll need to write data migrations for the model updates which aren't append-only. So if we decide to change model definitions throughout this process, it's fine but it's going to cause churn / time losses. In that spirit, I want to make sure we have a full handle on the specifics of the new model specs.

Note that we have ~three types of model changes

- add/delete model (no migration)
- append new properties (no migration)
- mutate definition (migration required)

## Initial questions

Firstly, during our sync tomorrow @ulaskozat I'd love to get a better semantic understanding of these changes, in case some of the data migrations detailed below can be dropped as unnecessary.

## Affected models, with questions

List of the directly-affected models, with any notes as necessary to indicate which parent model will actually need to be updated.

- qos_profile (mutate, make standalone network ent)
    - Notes
        - Was part of apn_configuration
    - ❓ App questions
        - Are all properties required (if not, what are default values)
        - Is id required to be non-empty string
        - Do the properties <priority_level, preemption_capability, preemption_vulnerability> from current model map to the relevant arp properties of the new model?
    - ❓ Migration questions
        - Should there be a default qos profile (per network)
- apn_configuration (mutate)
    - ❓ App questions
        - Are all properties required (if not, what are default values)
        - Is default_qos_profile required to be non-nil and/or non-empty
        - Is supported_qos_profiles required to be a superset of default_qos_profile
        - Is supported qos_profiles required to have non-nil and/or non-empty items
        - Why is it semantically required for non-default qos_profiles to have apn_ambr_dl/ul = 0
    - ❓ Migration questions
        - Should the existing per-network qos_profile property be migrated to a corresponding qos_profile ent
            - Copy over ambr values?
            - Copy over arp values?
            - What to fill for id, gbr, max_req_bw_dl, max_req_bw_ul properties of the "migrated" qos_profile
- apn_resource (new)
    - ❓ App questions
        - Are all properties required (if not, what are default values)
        - Should we do any crud-layer semantic validation? e.g. "uplink_bridge must be configured ..."
- lte_network (append)
- feg_lte_network (append)
- network_epc_configs (mutate)
    - Notes
        - Part of network_cellular_configs
    - ❓ Migration questions
        - What should the values for hss_* and gx_gy_* be if relay_enabled is true/false
- policy_rule (mutate)
    - ❓ Migration questions
        - What should the qos id be for migrated policy_rules
- flow_qos (delete)

## Required model changes

- Add
    - apn_resource
- Update
    - lte_network
    - feg_lte_network
- Updated, migration required
    - qos_profile
    - apn_configuration
    - network_cellular_configs (to mutate network_epc_configs)
    - policy_rule
- Delete
    - flow_qos

## Conclusion

Thoughts? E.g. @ulaskozat @andreilee

**Comment by andreilee** [comment]
@hcgatewood 
I updated the doc a little earlier today, fyi.

Hopefully this provides some additional details for now

- qos_profile
    - there will be two qos_profile types afterwards, one which is an apn_qos_profile, used for the apn_configuration, and is unchaged
    - there will be the policy_qos_profile assigned optionally to each policy
    - GBR and ARP are not required
- apn_configuration
    - unchaged by this proposal, read the updated doc
- apn_resource
    - if possible enforce the presence of ipv4_block and ipv6_prefix fields based on the pdn_type of the corresponding apn_configuration
- network_epc_configs
    - if relay_enabled is true, then the two new values are also true. hss_* and gx_gy_* together should act the same as relay_enabled

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 31]
```diff
@@ -0,0 +1,343 @@
...
+As before, a single QoS profile will be associated to an APN object at creation.
+This will define the QoS for the default bearer, in combination with the `ambr`
+field.
+
+```
+apn_resource:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN resource
+  ipv4_block: string
```
Comment (pshelar) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md]: We are not planning on supporting per APN IP block  in next release. It might be better to leave this configuration for later time.

**Review by ulaskozat — APPROVED** [review]

**Comment by andreilee** [comment]
Additional iterations will have new PRs opened, so that the changes are visible.

**Comment by hcgatewood** [comment]
# Orc8r clarifying questions v2

Okay went through the proposed orc8r changes. I have a few remaining questions (see below), and other than that just let me know if any of my assumptions seem off.

## Configurator DAG

Perhaps a bit unfortunately, configurator is intended to manipulate an acyclic graph. We haven't tested what functionality would break if we permitted cycles, and we don't have time to test it out and/or make changes during this sprint.

That means, in order to preserve the acyclic invariant of our network entity graph, we're going to have to flip the edge direction for at least some edges.

### Existing DAG

```
base_name -> (policy_rule ->) subscriber -> apn
```

### Proposed DAG

Adds the `apn_policy_profile`, to support a semantic `map[apn]policy_rule`

```
subscriber -.-> apn_policy_profile -.-> apn
            '-> apn                 '-> policy_rule*
            '-> policy_rule*
            '-> base_name -> policy_rule*

*policy_rule -> policy_qos_profile
```

### Resulting migration
- `base_name -> policy_rule` => `base_name <- policy_rule`
- `policy_rule -> subscriber` => `subscriber <- policy_rule`

## Model changes

### Affected models, with questions

- policy_qos_profile (new)
    - Notes
        - Correspond to existing policy_rule.qos (flow_qos)
        - Will "migrate" as part of policy_rule migration
    - ❓ App questions
        - Marshaling to FlowQos proto
            - Is class_id == FlowQos.Qci
            - If yes, do we want to enforce the FlowQos.Qci enum invariant
    - ❓ Migration questions
        - id := "%s_qos_profile".format(parent policy_rule.id)
        - max_req_bw_ul/dl := parent policy_rule.max_req_bw_ul/dl
        - What should the default values be for {gbr, arp}
- apn_resource (new)
- (mutable_)lte_gateway (append)
    - ❓ App questions
        - To confirm, we're specifically *not* updating/appending to the feg_lte_network model
- network_epc_configs (mutate)
    - Notes
        - Part of network_cellular_configs
        - Will set *_relay_enabled each to original value of relay_enabled
- (mutable_)subscriber (append)
    - Notes
        - See above dag notes
- policy_rule (mutate)
    - Notes
        - Will fully remove the flow_qos property, since migration is required anyway
        - Will flip the policy_rule->subscriber edge direction
        - See policy_qos_profile above for migration questions

### Required model changes

- Add
    - policy_qos_profile
    - apn_resource
    - apn_policy_profile (private)
- Update
    - lte_gateway, mutable_lte_gateway
    - subscriber, mutable_subscriber
- Update, migration required
    - network_cellular_configs (network_epc_configs property)
    - policy_rule
    - swap edge directions: base_name->policy_rule, base_name->subscriber

## Final questions

In the `base_name -> subscriber` relation, is a subscriber allowed to be assigned to multiple base_names?

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md [hunk 32]
```diff
@@ -0,0 +1,398 @@
...
+  # Keys present define the APNs that the subscriber has access to
+  # policy IDs define what policies will be installed for the subscriber for 
+  # each APN
+  active_policies_by_apn: dict[string, array[string]]
+  ...
+```
+
+To now specify APN-specific policies, the field `active_policies_by_apn` will
+be added. If an APN key is not present in `active_policies_by_apn`, then the
+subscriber does not have access to that APN.
```
Comment (ulaskozat) [inline docs/proposals/policy-feg-api-cleanup/qos_enforcement.md:243]: I missed this comment. So are we removing active_apns? How do we then enable subscriber access with default APN policy? Do we have a special name for it?

**Comment by andreilee** [comment]
@hcgatewood 

- policy_qos_profile
    - class_id is equal to qci
    - that id migration is fine
- feg_lte_network
    - not updating
- subscriber can be assigned multiple base names

File: docs/proposals/policy-feg-api-cleanup/qos_enforcement.md (hunk 32 above)
Comment (hcgatewood) [reply docs/proposals/policy-feg-api-cleanup/qos_enforcement.md:243]: Note for future: as resolved offline, this line should actually be "If an APN key is not present in `active_apns`, then the subscriber does not have access to that APN
